Feature/initial bring up intel ip repeater

adam1985d opened the pull request:
full description will be added here

PR for porting of Standalone Repeater from RDKB to PRPL and run over UGW8.x

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1040,6 +1039,9 @@ bool main_thread::backhaul_fsm_wireless(bool &skip_select)
         LOG_IF(!preferred_band_is_available, DEBUG) << "Preferred backhaul band is not available";
 
+        bool success        = true;
+        bool scan_trigerred = false;
```
Spelling mistake: scan_trigerred -> scan_triggered

adam1985d replied:
pushed a fixup commit. thanks

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1785,24 +1785,17 @@ bool main_thread::handle_1905_autoconfiguration_response(ieee1905_1::CmduMessage
 
         if (soc->sta_iface == m_sConfig.wireless_iface) {
-            notification->channel()   = iface_hal->get_channel();
-            notification->bandwidth() = radio_info.bandwidth;
-            notification->center_channel() =
```
Since bandwidth and center_channel are not valid, maybe they should be removed from the notification as well for now?

adam1985d replied:
these are stubs - and the next step is to utilize them to get full functionality.

at this stage the work done was to port the existing functionality in RDKB - but this change will be one of the first changes to be done to complete the full repeater functionality.

I leave it to you to decide if to remove them from the notification.

arnout commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+    #uci set beerocks.config.ssid='adam_wave_43'
+    ubus call uci set '{ "config" : "beerocks" ,"section" : "config", "values" : {"ssid" : "adam_GW_2.4G"}}'
+    #uci set beerocks.config.mode_enabled='None'
+    ubus call uci set '{ "config" : "beerocks" ,"section" : "config", "values" : {"mode_enabled" : "None"}}'
```
Better make a single ubus call that sets all values:

```
ubus call uci set '{ "config" : "beerocks" ,"section" : "config", 
    "values" : {
        "management_mode" : "Proprietary-Mesh",
        "operating_mode" : "WDS-Repeater",
        "ssid" : "adam_GW_2.4G",
        "mode_enabled" : "None"
    }}'
```

Same below, of course.

tomereli replied:
Why not simply use `uci set X=Y`???

adam1985d replied:
> Why not simply use `uci set X=Y`???

In UGW there is great emphasis on user permissions and ownership.
when you perform the uci set+commit  not via ubus you will change the ownership of the wireless/beerocks config file you are editing.
In the wireless uci configuration for example - this can prevent the hostapd from accessing the configuration file.
the owner of all configuration files uner /etc/config/ in UGW is rpcd and I followed the same line for beerocks

there is a discussion regarding UGW bringup and how exactly should this script and prplMesh/beerocks run and with which privileges - but decision wasn't made yet

tomereli replied:
OK. Anyway, the whole thing needs to be updated according to https://git.prpl.dev/prplmesh/iwlwav/blob/iwlwav-19.07/prplmesh/files/etc/config/prplmesh
As you can see the config name is no longer `beerocks`, it is `prplmesh`.

In addition, I think we should rename this file to `platform_intel.sh` which will be sourced by `prplmesh_utils.sh` and used if the platform is Intel (UGW or RDKB).

adam1985d replied:
I agree - but note there are 2 files for platform_daemon, one for RDKB (already available) and one for UGW (the new file introduced). we can check if unifying them to one file is beneficial - and how much they overlap or not as a future improvement

adam1985d replied:
changed the UCI set to one call for all params

arnout commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+                ;;
+        "start")
+                start
+                ;;
```
If there's a start, there should also be a stop.

adam1985d replied:
> If there's a start, there should also be a stop.

I have removed the start as the script should only configure the platform to repeater mode and not run the prplmesh which is a separate operation of the system/user

arnout commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+#!/bin/sh
```
Please add a copyright header.

adam1985d replied:
> Please add a copyright header.

done - please review it is as expected

arnout reviewed (commented):
I'm not very happy with including the platform_daemon_ugw.sh script. I think such scripts should be part of the platform integration (i.e. the openwrt package) because it's really related to that, more than to prplmesh itself.

However, for the time being, things are so much in flux that it's indeed better to include it here now. Just be warned that it will have to move out later :-)

tomereli commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+    
+    #uci commit wireless (currently automatically restarts the hostapd's)
+    ubus call uci commit '{"config" : "wireless" }'
+    # If the automatic restard is removed, the follwoing should be used after commit
```
Delete commented out lines

adam1985d replied:
done

tomereli commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+    /opt/prplMesh/scripts/prplmesh_utils.sh stop
+    
+    #(The daemon shouldn't perform this operation directly - but via UCI)
+    # Kill DHCP client
```
Delete commented out lines

adam1985d replied:
- this location is a stub to the operation that needs to be performed (with instruction to do it over UCI)
- I don't have the details of how to perform it over UCI yet and mention it needs to be part of this flow

I can remove but think it needs to be left here until edited with exact commands

tomereli replied:
You can leave it here as long as this PR is a draft, but if you'll want to merge it you'll have to clean it up.

tomereli commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+repeater_to_ap()
+{
+    # Kill APM
+    /opt/prplMesh/scripts/prplmesh_utils.sh stop
```
It is NOT APM. It is prplMesh. Also, no need to comment at all, its pretty clear what is done here.
Also, do not use absolute path in the script, the installation path might change. Better to call `/etc/init.d/prplmesh.sh` to perform the repeater to AP settings, stop and start prplmesh.

So:

```
/etc/init.d/prplmesh.sh repeater_mode
/etc/init.d/prplmesh.sh restart
```

All the rest should be deleted.

adam1985d replied:
so you mean to move the prpl (originally beerocks) uci configurations to the prplmesh.sh script?

tomereli replied:
yes, this is the platform script and is not part of prplmesh repository.
@arnout what do you think?

arnout replied:
I was the first one to comment that this script is a platform script and shouldn't be part of the prplmesh repository :-) Note that even openwrt-specific packages (ubus, uhttpd, ..) don't have their openwrt integration files as part of the package repository; they're always part of the openwrt repository.

Another advantage of moving this to the feed is that @blogic will review it. And he will (hopefully) make very appropriate comments like: this is the wrong place to set UCI defaults; that's something that should be done by a script in /etc/uci-defaults.

adam1985d replied:
as part of handling the bring-up on UGW that adds the support in init.d I will split the script functionality as requested here to:
1. repeater_mode configurations to prplmesh.sh
2. UCI defaults to the right location of UCI configurations (need to define what are defaults and what are according to mode of operation)

tomereli commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+    # Kill DHCP client
+	#kill `cat /var/run/udhcpc-br-lan.pid`
+
+    #configure APM
```
Not APM. Comment is meaningless, you should write what you are configuring if you already put a comment.

adam1985d replied:
removed

tomereli commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+	#kill `cat /var/run/udhcpc-br-lan.pid`
+
+    #configure APM
+    ubus call uci set '{ "config" : "beerocks" ,"section" : "config", "values" : {"management_mode" : "Multi-AP-Controller-and-Agent"}}'
```
Should go to repeater_mode in https://git.prpl.dev/prplmesh/iwlwav/blob/iwlwav-19.07/prplmesh/files/etc/init.d/prplmesh.sh (this script will be used in UGW as well).

adam1985d replied:
prplmesh.sh isn't available in init.d for UGW (only prplWrt)
will add it there after the add of prplmesh.sh to UGW init.d

tomereli commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+    #uci commit wireless (currently automatically restarts the hostapd's)
+    ubus call uci commit '{"config" : "wireless" }'
+    # If the automatic restard is removed, the follwoing should be used after commit
+    #/etc/init.d/network reload
```
Dead code - delete it.

adam1985d replied:
done

tomereli commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+    echo "Start APM"
+    /opt/prplMesh/scripts/prplmesh_utils.sh start
+}
+static_init()
```
what is `static_init()`? should add a function description. Also please add a line break before starting a new function.

tomereli commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+static_init()
+{
+    # Check if repeater mode (and required functionality is supported)
+    if [ $(uci get beerocks.config.management_mode) != "Proprietary-Mesh" ]
```
So `uci get` is fine but `uci set` has permissions issues in UGW? Why not switch to the correct user when running the script?

adam1985d replied:
we shouldn't run as rpcd - this is system user as far as I understand
the nwk user is more appropriate (but we might want to use a separate user - still in discussion)

uci get doesn't take ownership of the configuration file and thus has no side effects

tomereli commented on framework/platform/bpl/uci/scripts/platform_daemon_ugw.sh:
```diff
@@ -0,0 +1,146 @@
+{
+    # Check if repeater mode (and required functionality is supported)
+    if [ $(uci get beerocks.config.management_mode) != "Proprietary-Mesh" ]
+    then 
```
I'll ignore the rest, looks like still WIP

morantr commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1067,10 +1069,12 @@ bool main_thread::backhaul_fsm_wireless(bool &skip_select)
                 success = false;
                 break;
             }
+            scan_trigerred = true;
```
I think that if you will define the 'bool success = false` and here you will set it to true, you will get the same result.

adam1985d replied:
Thanks for your comment.

I want to differentiate between failure to perform the scan and unavailability of STAs to perform the scan (although both result in BH manager restart

morantr commented:
According to the https://github.com/prplfoundation/prplMesh/blob/master/CONTRIBUTING.md, max line length in commit message must be less than 72 charecters. This commits has exceeded it:
9695573c0a3f52b71eef07854e59cb35dee6e7b3
046809c4a2af272da4460b8fa78c13cef680ddf2
bb9ddcb9d4e7bb466225a0e3b39ef24d3dc1e3f0
d9a28541a045d2c1e9342380b1ebf91f32200874

spelling mistake "decribes"->"describes":
5589f0f4c084796900382bdde72f039452727ae2

adam1985d commented:
> We already have support to move to repeater mode in /etc/init.d/prplmesh.sh so what you did here should go there. And definitely not part of prplmesh repo, this has to go to the feed.

this is not yet supported for UGW

I will add a task to add init.d support and move the prpl configuration there - but this will a separate PR

arnout commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Initial BU Intel IP repeater to prplMesh](https://trello.com/c/eYDjEaBZ/32-initial-bu-intel-ip-repeater-to-prplmesh)

adam1985d committed:
agent: backhaul_fsm_wireless()::WPA_ATTACH - add missing break

In current implementation the attach would not break on failure and FSM
may change state to MASTER_DISCOVERY/INITIATE_SCAN depending only on
the "connected" condition instead of staying in WPA_ATTACH and retrying
to attach.

This change will allow continuous retries to attach.

As part of this commit the duplicated clear of "state_attempts" was
changes to a single location before checking the "connected" result.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
agent: backhaul_fsm_wireless() - read results only if scan triggered

As part of Repeater bring up for UGW8.x it was noticed that on the
event that no sta HAL is available, the wait for results is still
performed instead of restarting the backhaul FSM.

After the change, only if a scan is triggered as part of the
INITIATE_SCAN state, the FSM will move to WAIT_FOR_SCAN_RESULTS" state,
otherwise moves to the "RESTART" state.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
backhaul_manager: fix send_slaves_enable function

When RDKB code was ported to prplMesh the flow of Standalone Repeater
was broken by using refresh_radio_info() to get BW and center_channel,
but the refresh_radio_info() isn't doesn't support STAs yet.

The above-mentioned change is reverted and functionality will be
added only after the refresh_radio_info() will support STAs as well.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
framework: platform_daemon_ugw.sh - add platform script

As part of Standalone Repeater bring up in UGW it is required to
perform platform configuration before starting the APM.

The script added provides an easy way to perform these operations.

The script will be replaced by a platform_daemon that will handle
all such configurations and the script will guide the
platform_daemon implementation.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
ap_manager_thread::handle_hostapd_attached() - fix race bug

As part of Standalone Repeater bring up in RDKB we encountered failure
to perform read_acs_report() in the case that PRPL is started before
the ACS report is available.

After this change, the read_supported_channels() will be used which is
available even before the ACS scan was complete. The initial call to
read_acs_report() was to retrieve the supported_channels which is now
available using the read_supported_channels().

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
son_master:handle_intel_slave_join() - support IRE_MASTER

As part of Standalone Repeater bring up in RDKB there is a waver on
slave join flow - the condition of "parent-BSSID must be in connection
map before non-gw agent connects" was commented out.
The above-mentioned condition is required for MESH and RDKB did not
support MESH at the time of development.
If not commented out, the condition would fail since "far AP" - the AP
that the wireless backhaul connects to is not yet in the connection
map.

Commenting out the condition can break the MESH flow and thus the
waiver cannot be ported from RDKB.

To support both flows the condition is checked only if an agent
that joins is on a platform without a controller (MESH). In case a
controller is available on the platform the agent resides on, the
condition is not checked.

Signed-off-by: Adam Dov <adamx.dov@intel.com>
